include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -0,0 +1,342 @@
+    const int num_requested)
+{
+  const int n = t_maindiag.size();
+
+  assert(n >= 1);
+  assert(t_offdiag.size() == (size_t)(n - 1));
+
+  std::vector<double> evals;
+  std::vector<double> evecs; // flat array
+
+  if (n < num_requested)
+    return std::make_tuple(evals, evecs);
+
+  // Allocate storage
+  std::vector<double> matrix(n * n, 0.);
+
+  // Copy diagonals of the tridiagonal matrix into the full matrix
+  matrix[0] = t_maindiag[0];
+  for (int i = 1; i < n; ++i)
+  {
+    matrix[i + n * i] = t_maindiag[i];
+    matrix[i - 1 + n * i] = t_offdiag[i - 1];
+  }
+
+  // Eigenvalues are returned in ascending order
+  evals.resize(n);
+  const lapack_int info = LAPACKE_dsyev(LAPACK_COL_MAJOR, 'V', 'U', n,
+                                        matrix.data(), n, evals.data());
+  ASSERT(!info, "Call to LAPACKE_dsyev failed.");
+
+  // Save results.
+  evals.resize(num_requested);
+  evecs.resize(n * num_requested);
+  for (int i = 0; i < num_requested; ++i)
+  {
+    auto first = evecs.begin() + n * i;
+    auto t_first = matrix.begin() + n * i;
+    auto t_last = t_first + n;
+    double const norm =
+        std::sqrt(std::inner_product(t_first, t_last, t_first, 0.));
+    std::transform(t_first, t_last, first,
+                   [norm](auto &v) { return v / norm; });
+  }
+
+  return std::make_tuple(evals, evecs);
+}
+
+/// \brief Lanczos solver: perform convergence check
+template <typename OperatorType, typename VectorType>
+bool Lanczos<OperatorType, VectorType>::details_check_convergence(
+    double beta, const int num_evecs, const int num_requested, double tol,
+    std::vector<double> const &evecs)
+{
+  // Must iterate at least until we have num_requested eigenpairs
+  if (num_evecs < num_requested)
+    return false;
+
+  bool is_converged = true;
+
+  // Terminate if every approximage eigenvalue has converged to tolerance
+  // ISSUE: here ignoring possible nuances regarding the correctness
+  // of this check.
+  // NOTE: k may be desirable to "scale" the stopping criterion
+  // based on (estimate of) matrix norm or similar.
+  for (int i = 0; i < num_requested; ++i)
+  {
+    const double bound = beta * std::abs(evecs[num_evecs - 1 + num_evecs * i]);
+    is_converged = is_converged && bound <= tol;
+  }
+
+  return is_converged;
+}
+
+/// \brief Lanczos solver: calculate full (approx) eigenvectors from tridiag
+/// eigenvectors
+template <typename OperatorType, typename VectorType>
+std::vector<VectorType> Lanczos<OperatorType, VectorType>::details_calc_evecs(
+    const int num_requested, const int n,
+    std::vector<VectorType> const &lanc_vectors,
+    std::vector<double> const &evecs_tridiag)
+{
+  auto dim = lanc_vectors[0].size();
+
+  std::vector<VectorType> evecs(num_requested, VectorType(dim));
+
+  // Matrix-matrix product to convert tridiagonal eigenvectors to operator
+  // eigenvectors
+  for (int i = 0; i < num_requested; ++i)
+  {
+    evecs[i] = 0.0;
```

Comment:
Unnecessary since the vector is already initialized to zero.

Reply:
I was not sure if dealii does that. I would also like to leave it alone to make things clear.